src/gpb_parse.yrl

```diff
@@ -1059,60 +1059,91 @@ opt_tuple_to_atom_if_defined_true(Opt, Opts) ->
                  ({package,Name}) ->
-                      {package,maybe_tolower_name(Name,ToLower)};
+                      {package, maybe_tolower_or_snake_name(Name,ToLowerOrSnake)};
                  ({proto3_msgs,Names}) ->
                       {proto3_msgs,
-                       [prefix_suffix_name(Prefix, Suffix, ToLower, Name)
+                       [prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, Name)
                         || Name <- Names]};
                  (OtherElem) ->
                       OtherElem
               end,
               Defs).
 
-prefix_suffix_fields(Prefix, Suffix, ToLower, Fields) ->
+prefix_suffix_fields(Prefix, Suffix, ToLowerOrSnake, Fields) ->
     lists:map(
       fun(#?gpb_field{type={msg,MsgName}}=F) ->
-              NewMsgName = prefix_suffix_name(Prefix, Suffix, ToLower, MsgName),
+              NewMsgName = prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, MsgName),
               F#?gpb_field{type={msg,NewMsgName}};
+         (#?gpb_field{type={map,KeyType,{msg,MsgName}}}=F) ->
+              NewMsgName = prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, MsgName),
+              F#?gpb_field{type={map,KeyType,{msg,NewMsgName}}};
+         (#gpb_oneof{fields=Fs}=F) ->
+              Fs2 = prefix_suffix_fields(Prefix, Suffix, ToLowerOrSnake, Fs),
+              F#gpb_oneof{fields=Fs2};
          (#?gpb_field{}=F) ->
               F
       end,
       Fields).
 
-prefix_suffix_name(Prefix, Suffix, ToLower, Name) ->
-    Name1 = maybe_tolower_name(Name, ToLower),
-    Name2 = lists:concat([Prefix, Name1, Suffix]),
+prefix_suffix_name(Prefix, Suffix, ToLowerOrSnake, Name) ->
+    Name1 = maybe_tolower_or_snake_name(Name, ToLowerOrSnake),
+    Prefix1 = case Prefix of
+                  {by_proto, PrefixList} ->
+                      proplists:get_value(Name1, PrefixList, "");
+                  _ ->
+                      Prefix
+              end,
+    Name2 = lists:concat([Prefix1, Name1, Suffix]),
     list_to_atom(Name2).
 
-maybe_tolower_name(Name, false) -> Name;
-maybe_tolower_name(Name, true) ->
-    list_to_atom(string:to_lower(atom_to_list(Name))).
-
-prefix_suffix_rpcs(Prefix, Suffix, ToLower, RPCs) ->
+maybe_tolower_or_snake_name(Name, false) -> Name;
+maybe_tolower_or_snake_name(Name, to_lower) ->
+    list_to_atom(string:to_lower(atom_to_list(Name)));
+maybe_tolower_or_snake_name(Name, snake_case) ->
+    NameString = atom_to_list(Name),
+    Snaked = lists:foldl(fun(RE, Snaking) ->
+                             re:replace(Snaking, RE, "\\1_\\2", [{return, list}, global])
+                         end, NameString, ["(.)([A-Z][a-z]+)",    %% uppercase followed by lowercase, except beginning
+                                           "(.)([0-9]+)",         %% any consecutive digits, except beginning
+                                           "([a-z0-9])([A-Z])"]), %% uppercase with lowercase or digit before it
```

Comment:
Wow, this regexp actually works (well, of course it does)
I'm dazed and amazed.
Tried it with a few examples, and it appears to do what I'd expect it to do.